### PR TITLE
fix: avoid inadvertent deletion of active HSM keys

### PR DESCRIPTION
**Pull request by nklaassen**
This is a partial fix for #25017

The latest version of the YubiHSM2 SDK has changed the behavior for keys longer than 2 bytes, which used to be silently truncated for all operations.
This causes an unfortunate interaction with `DeleteUnusedKeys` when the SDK is upgraded in an active Teleport cluster.
Because none of the active keys can be queried from the HSM individually by their ID, but they can be listed by their label, all of the active keys end up being deleted.

`DeleteUnusedKeys` is written this way in an attempt to be "stateless". Trying to synchronously delete keys at the instant they are rotated out during a CA rotation would be error-prone. If the auth server were to restart or crash at the wrong moment, you could be left with an orphaned key on your HSM forever, with no reference to it stored by Teleport or anywhere else.

Instead, the Auth server labels all keys it creates with its own host UUID.
Then periodically (during startup) it lists all keys in the HSM that are labeled with its own UUID, and if they are not currently active, deletes them.
This goes catastrophically wrong when individual lookup operations fail, but list operations succeed.

The fix here is to avoid deleting any keys if any single lookup fails.

The YubiHSM2 SDK version `2023.1` is still not supported, but with this fix at least we won't delete any active keys.

### CODE REVIEW

File: lib/auth/keystore/pkcs11.go [hunk 1]
```diff
@@ -212,37 +212,53 @@ func (p *pkcs11KeyStore) deleteKey(_ context.Context, rawKey []byte) error {
...
+// rotation.
+func (p *pkcs11KeyStore) DeleteUnusedKeys(ctx context.Context, activeKeys [][]byte) error {
 	p.log.Debug("Deleting unused keys from HSM")
-	var usedPublicKeys []*rsa.PublicKey
-	for _, usedKey := range usedKeys {
-		if keyType(usedKey) != types.PrivateKeyType_PKCS11 {
+
+	// It's necessary to fetch all PublicKeys for the known activeKeys in order to
+	// compare with the signers returned by FindKeyPairs below. We have no way
+	// to find the CKA_ID of an unused key if it not known.
```
Comment (AntonAM) [inline lib/auth/keystore/pkcs11.go:224]:
```suggestion
	// to find the CKA_ID of an unused key if it is not known.
```

**Review by AntonAM — APPROVED** [review]
Yeah, pretty unfortunate. How did we found out, from an angry customer? 😅

**Comment by nklaassen** [comment]
Luckily I found it while running through the test plan before a customer found it for us 😬 I have reached out to the one customer who I know is using this hardware to warn them about the issue, I will feel a bit better once we ship this fix

File: lib/auth/keystore/gcp_kms.go [hunk 2]
```diff
@@ -263,6 +269,14 @@ func (g *gcpKMSKeyStore) DeleteUnusedKeys(ctx context.Context, usedKeys [][]byte
 		return trace.Wrap(err)
 	}
 
+	for _, found := range activeKmsKeyVersions {
+		if found == 0 {
+			// Failed to find a currently active key owned by this host.
+			// The cluster is in a bad state, refuse to delete any keys.
+			return trace.NotFound("Failed to find currently active CA key in GCP KMS, aborting attempt to delete unused keys")
```
Comment (codingllama) [inline lib/auth/keystore/gcp_kms.go:276]: nit: lowercase error, avoid words like "failed"

```suggestion
			return trace.NotFound("cannot find currently active CA key in GCP KMS, aborting attempt to delete unused keys")
```

Also, do you think it's worth returning or logging some sort of key identifier for the "bad" keys?

File: lib/auth/keystore/pkcs11.go [hunk 3]
```diff
@@ -212,37 +212,53 @@ func (p *pkcs11KeyStore) deleteKey(_ context.Context, rawKey []byte) error {
...
+			// This key was labeled with a foreign host UUID, it is likely not
+			// present on the attached HSM and definitely will not be returned
+			// by FindKeyPairs below which queries by host UUID.
 			continue
 		}
+		signer, err := p.getSigner(ctx, activeKey)
+		if trace.IsNotFound(err) {
+			// Failed to find a currently active key owned by this host.
+			// The cluster is in a bad state, refuse to delete any keys.
+			return trace.NotFound("Failed to find currently active CA key in HSM, aborting attempt to delete unused keys")
```
Comment (codingllama) [inline lib/auth/keystore/pkcs11.go:244]:
```suggestion
			return trace.NotFound("cannot find currently active CA key in HSM, aborting attempt to delete unused keys")
```

File: lib/auth/keystore/pkcs11.go [hunk 4]
```diff
@@ -254,12 +270,13 @@ func (p *pkcs11KeyStore) DeleteUnusedKeys(ctx context.Context, usedKeys [][]byte
...
+		if keyIsActive(signer) {
 			continue
 		}
 		if err := signer.Delete(); err != nil {
-			// Key deletion is best-effort, log a warning on errors. Errors have
-			// been observed when FindKeyPairs returns duplicate keys.
+			// Key deletion is best-effort, log a warning on errors, and
+			// continue trying to delete other keys. Errors have been observed
+			// when FindKeyPairs returns duplicate keys.
 			p.log.Warnf("failed deleting unused key from HSM: %v", err)
```
Comment (codingllama) [inline lib/auth/keystore/pkcs11.go:280]:
```suggestion
			p.log.Warnf("Failed deleting unused key from HSM: %v", err)
```

**Review by codingllama — APPROVED** [review]

**Pull request merged**